overlays/haskell.nix

```
@@ -541,6 +541,50 @@ final: prev: {
               shells.ghc = p.hsPkgs.shellFor {};
             };
 
+        # `project'` and `project` automatically select between `cabalProject`
+        # and `stackProject` (when possible) by looking for `stack.yaml` or
+        # `cabal.project` files.  If both exist we can pass in one of:
+        #     `projectFileName = "stack.yaml;"`
+        #     `projectFileName = "cabal.project";`
+        # to let it know which to choose (or pick another name).  If the
+        # selected file ends in a `.yaml` it is assumed to be for `stackProject`.
+        # If niether `stack.yaml` nor `cabal.project` exist `cabalProject` is
+        # used (as it will use a default `cabal.project`).
+        project' = { src, projectFileName ? null, ... }@args: 
+          let
+            dir = __readDir (src.origSrcSubDir or src);
+            exists = fileName: builtins.elem (dir.${fileName} or "") ["regular" "symlink"];
+            stackYamlExists    = exists "stack.yaml";
+            cabalProjectExists = exists "cabal.project";
+            selectedFileName =
+              if projectFileName != null
+                then projectFileName  # Prefer the user selected project file name
+                else
+                  if stackYamlExists && cabalProjectExists
+                    then null # If both exist we will throw an error
+                    else
+                      if stackYamlExists
+                        then "stack.yaml"      # stack needs a stack.yaml
+                        else "cabal.project";  # the cabal.project file is optional
```

I'm not so happy about the `null` usage here.
```suggestion
                  if stackYamlExists
                    then "stack.yaml"      # stack needs a stack.yaml
                    else "cabal.project";  # the cabal.project file is optional
```